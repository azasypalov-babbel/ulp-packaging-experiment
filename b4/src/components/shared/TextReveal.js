import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Text from './Text';

const StyledTextReveal = styled.div`
  box-sizing: border-box;
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100%;

  background-color: ${(props) =>
    props.reveal
      ? props.theme.cascada.iceGray
      : props.theme.cascada.pastelElectric
};

  transition: background-color 0.5s;

  max-width: 20rem;

  border-radius: 0.25rem;

  padding: 1rem;
`;

const TextReveal = ({ reveal, children }) =>(
  <StyledTextReveal reveal={reveal}>
    <Text
      fontFamily="fontMilliardMedium"
      color="spaceGray"
      fontSize={reveal ? 'medium' : 'large'}
      textAlign="center">
      {reveal ? children : '?'}
    </Text>
  </StyledTextReveal>
);

TextReveal.propTypes = {
  reveal: PropTypes.bool,
  children: PropTypes.node
};

export default TextReveal;
