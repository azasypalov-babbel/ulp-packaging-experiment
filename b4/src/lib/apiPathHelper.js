export const buildPath = ({ version, locale, uuid, learnLanguageAlpha3 }, rest = '') => {
  if (!version) {
    throw 'Missing required param "version"';
  }
  if (!locale) {
    throw 'Missing required param "locale"';
  }
  if (!uuid) {
    throw 'Missing required param "uuid"';
  }
  if (!learnLanguageAlpha3) {
    throw 'Missing required param "learnLanguageAlpha3"';
  }

  return `v${version}/${locale}/accounts/${uuid}/learn_languages/${learnLanguageAlpha3}/${rest}`;
};

export const buildPathV5 = ({ version, locale, uuid, contentReleaseId, learnLanguageAlpha3 }, rest = '') => {
  if (!version) {
    throw 'Missing required param "version"';
  }
  if (!locale) {
    throw 'Missing required param "locale"';
  }
  if (!uuid) {
    throw 'Missing required param "uuid"';
  }
  if (!learnLanguageAlpha3) {
    throw 'Missing required param "learnLanguageAlpha3"';
  }
  if (!contentReleaseId) {
    throw 'Missing required param "contentReleaseId"';
  }

  return `v${version}/${locale}/accounts/${uuid}/content_releases/${contentReleaseId}` +
    `/learn_languages/${learnLanguageAlpha3}/${rest}`;
};
