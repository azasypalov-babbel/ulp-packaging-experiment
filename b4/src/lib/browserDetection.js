// please note,
// that IE11 now returns undefined again for window.chrome
// and new Opera 30 outputs true for window.chrome
// but needs to check if window.opr is not undefined
// and new IE Edge outputs to true now for window.chrome
// and if not iOS Chrome check
// so use the below updated condition

// based on https://stackoverflow.com/a/13348618
const isChromium = window.chrome;
const agent = window.navigator.userAgent.toLowerCase();
const vendorName = window.navigator.vendor;
const isOpera = typeof window.opr !== 'undefined';
const isIEedge = agent.indexOf('edge') > -1 || agent.indexOf('edg') > -1;
let isChrome = false;

if (
  isChromium !== null &&
  typeof isChromium !== 'undefined' &&
  vendorName === 'Google Inc.' &&
  isOpera === false &&
  isIEedge === false
) {
  isChrome = true;
}

export { isChrome };
