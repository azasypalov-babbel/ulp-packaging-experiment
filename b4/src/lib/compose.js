export const compose = (...functionsToCompose) => (argumentsOfFunctions) => {
  return functionsToCompose.reduceRight((result, functionToCompose) => {
    return functionToCompose(result);
  }, argumentsOfFunctions);
};
