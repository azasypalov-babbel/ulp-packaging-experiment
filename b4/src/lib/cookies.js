function identity(x) {
  return x;
}

function convertFeatureStrToObject(elem) {
  const elements = elem.trim().split('=');
  return {
    [elements[0]]: elements[1]
  };
}

function arrayToObjectReducer(acc, element) {
  return {
    ...acc,
    ...element
  };
}

function getCookies() {
  return document?.cookie.split(';')
    .filter(identity)
    .map(convertFeatureStrToObject)
    .reduce(arrayToObjectReducer, {});
}

function getCookie(cookie) {
  const cookies = getCookies();

  return cookies && cookies[cookie];
}

function setCookie(key, value, expiryDate) {
  document.cookie = `${key}=${value};expires=${expiryDate.toUTCString()};path=/`;
}

export default {
  setCookie,
  getCookies,
  getCookie
};
