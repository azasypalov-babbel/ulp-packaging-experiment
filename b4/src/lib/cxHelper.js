import baseCX from 'classnames';

const PREFIX = 'loy';

const cx = (...args) => {
  const classNames = baseCX(args);

  return classNames.split(' ')
    .map((className) => `${PREFIX}-${className}`)
    .join(' ');
};

export default cx;
