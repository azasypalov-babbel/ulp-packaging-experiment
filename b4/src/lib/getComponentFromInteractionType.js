import Fillin from '../components/Interactions/Fillin/FillinInteraction';
import Puzzlehelper from '../components/Interactions/Puzzlehelper/PuzzlehelperInteraction';
import Wordorder from '../components/Interactions/Wordorder/WordorderInteraction';
import ChoiceButtons from '../components/Interactions/Choicebuttons/ChoicebuttonsInteraction';
import Speak from '../components/Interactions/Speak/SpeakInteraction';

const interactionMap = {
  speak: Speak,
  fillin: Fillin,
  puzzlehelper: Puzzlehelper,
  wordorder: Wordorder,
  choose: ChoiceButtons
};

const getComponentFromInteractionType = (interactionType) => interactionMap[interactionType];

export default getComponentFromInteractionType;

