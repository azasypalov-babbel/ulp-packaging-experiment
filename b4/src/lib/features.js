import cookies from './cookies';
const { setCookie, getCookies } = cookies;

let featureStore = {};

const defaultAllowedFeatures = [
  'is_demo',
  'is_redux_logger',
  'is_redux_devtools',
  'is_unlocked',
  'is_cat_preview',
  'is_analyzer_lib_recognizer',
  'is_b3_vocabulary_trainer',
  'is_web_speech_mock',
  'is_native_speech',
  'is_verbose_logging',
  'is_transliteration',
  'is_fake_caret',
  'is_selection_helper',
  'is_puzzle_helper',
  'is_debug',
  'is_refer_a_friend',
  'is_zendesk'
];

const defaultNonPersistentFeatures = [
  'is_demo',
  'is_analyzer_lib_recognizer',
  'is_b3_vocabulary_trainer',
  'is_fake_caret',
  'is_selection_helper',
  'is_puzzle_helper',
  'is_zendesk'
];

function filterFeatures(features, allowedFeatures) {
  const filteredFeatures = {};
  Object.keys(features).forEach(function(key) {
    if (allowedFeatures.includes(key) && (features[key] === 'on' || features[key] === 'off')) {
      filteredFeatures[key] = features[key];
    }
  });
  return filteredFeatures;
}

function filterPersistentFeatures(features, nonPersistentFeatures) {
  const filteredFeatures = {};
  Object.keys(features).forEach(function(key) {
    if (!nonPersistentFeatures.includes(key)) {
      filteredFeatures[key] = features[key];
    }
  });
  return filteredFeatures;
}

export function configure(
  features,
  allowedFeatures = defaultAllowedFeatures,
  nonPersistentFeatures = defaultNonPersistentFeatures
) {
  featureStore = filterFeatures(features, allowedFeatures);

  const persistantFeatures = filterPersistentFeatures(
    filterFeatures(getCookies(), allowedFeatures),
    nonPersistentFeatures
  );

  Object.entries(persistantFeatures)
    .forEach(function([key, value]) {
      if (!(key in featureStore)) {
        featureStore[key] = value;
      }
    });

  const expiryDate = new Date();
  const oneYearMs = 365 * 24 * 60 * 60 * 1000;
  expiryDate.setTime(expiryDate.getTime() + oneYearMs);

  Object.entries(filterPersistentFeatures(featureStore, nonPersistentFeatures))
    .forEach(function([key, value]) {
      setCookie(key, value, expiryDate);
    });
}

export function get(featureName) {
  if (featureStore[featureName] === 'on') {
    return true;
  } else if (featureStore[featureName] === 'off') {
    return false;
  } else {
    return null;
  }
}

export function getActiveFeatures() {
  return Object.keys(featureStore)
    .filter((key) => featureStore[key] === 'on');
}

export function isWebview() {
  return process.env.WEBVIEW === '1';
}

export function isAndroid() {
  return (
    isWebview()
    && typeof window.androidListener?.postMessage === 'function'
  );
}

export function isiOS() {
  return (
    isWebview() &&
    typeof window.webkit?.messageHandlers?.iosListener?.postMessage === 'function'
  );
}
