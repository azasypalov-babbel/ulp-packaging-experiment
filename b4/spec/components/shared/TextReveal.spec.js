import React from 'react';
import TextReveal from '../../../src/components/shared/TextReveal';
import { mountWithTheme } from './themeMock';
import { isWebview } from '../../../src/lib/features';
jest.mock('../../../src/lib/features');

describe('<TextReveal />', () => {
  describe('when NOT in Webview', () => {
    beforeEach(() => {
      isWebview.mockImplementationOnce(() => false);
    });

    test('it renders', () => {
      const wrapper = mountWithTheme(<TextReveal />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('when in Webview', () => {
    beforeEach(() => {
      isWebview.mockImplementationOnce(() => true);
    });

    test('it renders', () => {
      const wrapper = mountWithTheme(<TextReveal />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
