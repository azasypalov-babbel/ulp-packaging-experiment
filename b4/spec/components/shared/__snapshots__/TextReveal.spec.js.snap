// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextReveal /> when NOT in Webview it renders 1`] = `
.c1 {
  font-family: Milliard-Medium,sans-serif;
  color: #394149;
  font-weight: normal;
  line-height: 1.25;
  -webkit-letter-spacing: 0.031rem;
  -moz-letter-spacing: 0.031rem;
  -ms-letter-spacing: 0.031rem;
  letter-spacing: 0.031rem;
  text-align: center;
}

.c1 > strong,
.c1 > b {
  font-family: Milliard-SemiBold,sans-serif;
  font-weight: normal;
}

.c1 > em,
.c1 > i {
  font-family: Milliard-BookItalic,sans-serif;
  font-weight: normal;
  font-style: normal;
}

.c1 > strong em,
.c1 > b i {
  font-family: Milliard-MediumItalic,sans-serif;
  font-weight: normal;
}

.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 100%;
  background-color: #d1eefc;
  -webkit-transition: background-color 0.5s;
  transition: background-color 0.5s;
  max-width: 20rem;
  border-radius: 0.25rem;
  padding: 1rem;
}

<TextReveal>
  <styled.div>
    <div
      className="c0"
    >
      <styled.span
        color="spaceGray"
        fontFamily="fontMilliardMedium"
        fontSize="large"
        textAlign="center"
      >
        <span
          className="c1"
          color="spaceGray"
          fontFamily="fontMilliardMedium"
          fontSize="large"
        >
          ?
        </span>
      </styled.span>
    </div>
  </styled.div>
</TextReveal>
`;

exports[`<TextReveal /> when in Webview it renders 1`] = `
.c1 {
  font-family: Milliard-Medium,sans-serif;
  color: #394149;
  font-weight: normal;
  line-height: 1.25;
  -webkit-letter-spacing: 0.031rem;
  -moz-letter-spacing: 0.031rem;
  -ms-letter-spacing: 0.031rem;
  letter-spacing: 0.031rem;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 > strong,
.c1 > b {
  font-family: Milliard-SemiBold,sans-serif;
  font-weight: normal;
}

.c1 > em,
.c1 > i {
  font-family: Milliard-BookItalic,sans-serif;
  font-weight: normal;
  font-style: normal;
}

.c1 > strong em,
.c1 > b i {
  font-family: Milliard-MediumItalic,sans-serif;
  font-weight: normal;
}

.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 100%;
  background-color: #d1eefc;
  -webkit-transition: background-color 0.5s;
  transition: background-color 0.5s;
  max-width: 20rem;
  border-radius: 0.25rem;
  padding: 1rem;
}

<TextReveal>
  <styled.div>
    <div
      className="c0"
    >
      <styled.span
        color="spaceGray"
        fontFamily="fontMilliardMedium"
        fontSize="large"
        textAlign="center"
      >
        <span
          className="c1"
          color="spaceGray"
          fontFamily="fontMilliardMedium"
          fontSize="large"
        >
          ?
        </span>
      </styled.span>
    </div>
  </styled.div>
</TextReveal>
`;
